text/3573-is.md

```diff
@@ -0,0 +1,284 @@
+`let Some(y) = x && y > 5`.
+
+This is even more true at the end of a longer expression chain, such as
+`x.method()?.another_method().await? is Some(y)`. Rust method chaining and `?`
+and `.await` all encourage writing code that reads in operation order from left
+to right, and `is` fits naturally at the end of such a sequence.
+
+Having an `is` operator would also help to reduce the proliferation of methods
+on types such as `Option` and `Result`, by allowing prospective users of those
+methods to write a condition using `is` instead. While any such condition could
+equivalently be expressed using `let`-chains, the binding would then move
+further away from the condition expression referencing the binding, which would
+result in a less natural reading order for the expression.
+
+Consider the following examples:
+
+```rust
+if expr_producing_option().is_some_and(|v| condition(v))
+
+if let Some(v) = expr_producing_option() && condition(v)
+
+if expr_producing_option() is Some(v) && condition(v)
+```
+
+The condition using `is` is a natural translation from the `is_some_and`
+method, whereas the if-let construction requires reversing the binding of `v`
+and the expression producing the option. This seems sufficiently cumbersome in
+some cases that the absence of `is` would motivate continued use and
+development of helper methods.
+
+# Guide-level explanation
+
+Rust provides an `is` operator, which can be used in any expression:
+`EXPR is PATTERN`
+
+This operator tests if the value of `EXPR` matches the specified `PATTERN`; see
+<https://doc.rust-lang.org/reference/patterns.html> for details on patterns.
+
+If the `EXPR` matches the `PATTERN`, the `is` expression evaluates to `true`,
+and additionally binds any bindings specified in `PATTERN` in the current scope
+for code subsequently executed along the path where the `is` expression
+is known to be `true`.
+
+For example:
+
+```rust
+if an_option is Some(x) && x > 3 {
+    println!("{x}");
+}
+```
+
+The bindings in the pattern are not bound along any code path potentially
+reachable where the expression did not match:
+
+```rust
+if (an_option is Some(x) && x > 3) || (more_conditions /* x is not bound here*/) {
+    // x is not bound here
+} else {
+    // x is not bound here
+}
+// x is not bound here
+```
+
+The pattern may use alternation (within parentheses), but must have the same
+bindings in every alternative:
+
+```rust
+if color is (RGB(r, g, b) | RGBA(r, g, b, _)) && r == b && g < 10 {
+    println!("condition met")
+}
+
+// ERROR: `a` is not bound in all alternatives of the pattern
+if color is (RGB(r, g, b) | RGBA(r, g, b, a)) && r == b && g < 10 {
+    println!("condition met")
+}
+```
+
+`is` may appear anywhere a boolean expression is accepted:
+
+```rust
+func(x is Some(y) && y > 3);
+```
+
+The `is` operator may not appear as a statement; use `let` to bind a pattern in
+that context:
+
+```rust
+// ERROR: use `let` for this
+an_expression() is x;
```

Comment:
This seems like it should be a deny-by-default lint, not a hard error, and the reasoning is a bit confusing here.

To me, this is just a special case of an unused expression: since `x` is only bound for the expression itself, it doesn't get used beyond the expression.

Like, I mostly agree with what you're trying to explain here, but I think that we should be careful to explain the true reason why this is wrong, rather than leaning into misconceptions harder in an attempt to debunk them.

Reply:
I agree, this makes sense as a deny-by-default lint, and I can provide a more detailed explanation that the binding would not be usable after the `;`.